dialogue-core/src/main/java/com/palantir/dialogue/core/Balanced.java

```diff
@@ -0,0 +1,172 @@
+     */
+    @VisibleForTesting
+    static final Comparator<ChannelWithStats> BY_SCORE = Comparator.comparing(ChannelWithStats::score);
+
+    private final ImmutableList<ChannelWithStats> channels;
+    private final Random random;
+    private final CodahaleClock clock;
+
+    Balanced(ImmutableList<LimitedChannel> channels, Random random, Ticker ticker) {
+        Preconditions.checkState(channels.size() >= 2, "At least two channels required");
+        this.random = random;
+        this.clock = new CodahaleClock(ticker);
+        this.channels = channels.stream()
+                .map(channel -> new ChannelWithStats(channel, clock))
+                .collect(ImmutableList.toImmutableList());
+
+        log.debug("Initialized", SafeArg.of("count", channels.size()), UnsafeArg.of("channels", channels));
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        if (log.isDebugEnabled()) {
+            log.debug(
+                    "Channel scores {}",
+                    SafeArg.of("scores", Lists.transform(channels, ChannelWithStats::score)),
+                    SafeArg.of("inflight", Lists.transform(channels, ChannelWithStats::inflight)));
+        }
+
+        // pre-shuffling is pretty important here, otherwise when there are no requests in flight, we'd
+        // *always* prefer the first channel of the list, leading to a higher overall load.
+        List<ChannelWithStats> preShuffled = shuffleImmutableList(channels, random);
+
+        // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
+        // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
+        return preShuffled.stream()
+                .sorted(BY_SCORE)
+                .map(channel -> channel.maybeExecute(endpoint, request))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                // we rely heavily on streams being lazy here
+                .findFirst();
+    }
+
+    /** Returns a new shuffled list, without mutating the input list (which may be immutable). */
+    private static <T> List<T> shuffleImmutableList(ImmutableList<T> sourceList, Random random) {
+        List<T> mutableList = new ArrayList<>(sourceList);
+        Collections.shuffle(mutableList, random);
+        return mutableList;
+    }
+
+    static final class ChannelWithStats implements LimitedChannel {
+        private final LimitedChannel delegate;
+
+        @VisibleForTesting
+        final AtomicInteger inflight = new AtomicInteger(0);
+
+        /**
+         * We keep track of failures within a time window to do well in scenaios where an unhealthy server returns
+         * errors much faster than healthy nodes can serve good responses. See
+         * {@link SimulationTest#fast_503s_then_revert}.
+         */
+        @VisibleForTesting
+        final SlidingTimeWindowArrayReservoir recentFailures;
```

Comment:
Have we validated the implementation of reservoir for read performance? They're generally written to be performant with more reads than writes, but our usage pattern is the opposite.

Reply:
Not yet - I think this one would end up calling `ChunkedAssociativeLongArray#size`, which doesn't look _too_ horrible...

```java
   synchronized int size() {
        int result = 0;
        for (Chunk chunk : chunks) {
            result += chunk.cursor - chunk.startIndex;
        }
        return result;
    }
```

Reply:
Since full accuracy isn't really important here, you can use something simpler but a bit more fuzzy. Pseudocode:

```
class FuzzyReservoir {
    long count;
    long[] windows;
    int cursor;
    long rolloverNanos;
    long nextRollover;

    FuzzerReservoir(long sizeNanos, int windows) {
        windows = new long[windows];
        rolloverNanos = sizeNanos / windows;
        nextRollover = System.nanoTime() + sizeNanos;
    }


    void increment() {
        maybeRoll();
        count += 1;
        windows[cursor] += 1;
    }

    long get() {
        return count;
    }

    void maybeRoll() {
        if (System.nanoTime() < nextRollover) {
            return;
        }

        cursor = (cursor + 1) % windows.length;
        count -= windows[cursor];
        windows[cursor] = 0;
        nextRollover += rolloverNanos;
    }
}
```

If you make count an atomic, you can have lock free reads, and just synchronize writes.

Reply:
Generally in favor of the idea, we would need to check for rollover on read otherwise a high-cost channel will never be selected, and have no way to decay cost over time.

Reply:
Have implemented, see LockFreeTimeWindowReservoir.java